src/main/java/dev/morling/onebrc/CalculateAverage_berry120.java

```diff
@@ -0,0 +1,256 @@
+            return "TemperatureSummary{" +
+                    "name=" + new String(name) +
+                    ", min=" + min +
+                    ", max=" + max +
+                    ", total=" + total +
+                    ", sampleCount=" + sampleCount +
+                    '}';
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+
+        // long time = System.currentTimeMillis();
+
+        Path path = Path.of(FILE);
+        RandomAccessFile file = new RandomAccessFile(path.toFile(), "r");
+        FileChannel channel = file.getChannel();
+        long size = Files.size(path);
+        int splitSize = size < 10_000_000 ? 1 : SPLIT_SIZE;
+        long inc = (int) (size / splitSize);
+
+        List<Long> positions = new ArrayList<>();
+        positions.add(0L);
+
+        long pos = 0;
+        for (int i = 0; i < splitSize; i++) {
+            long startPos = pos;
+            long endPos = pos + inc;
+            MappedByteBuffer bb = channel.map(FileChannel.MapMode.READ_ONLY, startPos, endPos - startPos);
+            int dec = 1;
+            while (bb.get((int) inc - dec) != '\n') {
+                dec++;
+            }
+            endPos -= (dec - 1);
+            pos = endPos;
+            positions.add(pos);
+        }
+        positions.add(size);
+
+        List<TemperatureSummary[]> arrs = Collections.synchronizedList(new ArrayList<>());
+
+        var tp = Executors.newFixedThreadPool(splitSize);
+
+        // System.out.println("SETUP TIME: " + (System.currentTimeMillis() - time));
+
+        for (int i = 1; i < positions.size(); i++) {
+            final int idx = i;
+            tp.submit(() -> {
+                TemperatureSummary[] arr = new TemperatureSummary[0xFFFFF];
+                arrs.add(arr);
+                try {
+                    long startPos = positions.get(idx - 1);
+                    long endPos = positions.get(idx);
+                    int bSize = (int) (endPos - startPos);
+                    MappedByteBuffer bb = channel.map(FileChannel.MapMode.READ_ONLY, startPos, bSize);
+
+                    byte[] bytes = new byte[0xFFFFF];
+                    int bytesDataLength;
+
+                    int breakIdx = -1;
+                    int rollingHash = 5381;
+                    int sourceIdx = 0;
+                    int startReadingIdx = 0;
+
+                    bb.get(sourceIdx, bytes, 0, bytesDataLength = Math.min(bytes.length, bSize));
+
+                    int bi = 0;
+                    while (true) {
+
+                        byte b = bytes[bi];
+
+                        if (b == ';') {
+                            breakIdx = bi + 1;
+                            bi += 4;
+                        }
+                        else if (breakIdx == -1) {
+                            // 1M constrained djb2 should be good enough for our purposes
+                            rollingHash = (((rollingHash << 5) + rollingHash) + b) & 0xFFFFF;
```

Comment:
Hum, seems this would not reliably prevent hash collisions? If so, it would not be compliant ("may not make assumptions about the data set"). You'd have to implement some means of open addressing, unless I'm missing something.

Reply:
Fixed 👍